Source/FieldSolver/SpectralSolver/SpectralAlgorithms/PsatdAlgorithmFirstOrder.cpp

```diff
@@ -0,0 +1,368 @@
+            const Complex Jz_new = (J_linear  ) ? fields(i,j,k,Idx.Jz_new) : 0._rt;
+
+            const Complex Jx_c0 = (J_constant) ? Jx_mid : Jx_old;
+            const Complex Jy_c0 = (J_constant) ? Jy_mid : Jy_old;
+            const Complex Jz_c0 = (J_constant) ? Jz_mid : Jz_old;
+            const Complex Jx_c1 = (J_linear  ) ? (Jx_new-Jx_old)/dt : 0._rt;
+            const Complex Jy_c1 = (J_linear  ) ? (Jy_new-Jy_old)/dt : 0._rt;
+            const Complex Jz_c1 = (J_linear  ) ? (Jz_new-Jz_old)/dt : 0._rt;
+
+            Complex rho_mid, rho_old, rho_new, F_old, G_old;
+            Complex rho_c0, rho_c1;
+            if (div_cleaning)
+            {
+                rho_mid = (rho_constant) ? fields(i,j,k,Idx.rho_mid) : 0._rt;
+                rho_old = (rho_linear  ) ? fields(i,j,k,Idx.rho_old) : 0._rt;
+                rho_new = (rho_linear  ) ? fields(i,j,k,Idx.rho_new) : 0._rt;
+
+                F_old = fields(i,j,k,Idx.F);
+                G_old = fields(i,j,k,Idx.G);
+
+                rho_c0 = (rho_constant) ? rho_mid : rho_old;
+                rho_c1 = (rho_linear  ) ? (rho_new-rho_old)/dt : 0._rt;
+            }
+
+            // k vector values
+            const amrex::Real kx = modified_kx_arr[i];
+#if defined(WARPX_DIM_3D)
+            const amrex::Real ky = modified_ky_arr[j];
+            const amrex::Real kz = modified_kz_arr[k];
+#else
+            constexpr amrex::Real ky = 0._rt;
+            const     amrex::Real kz = modified_kz_arr[j];
+#endif
+            // Physical constants and imaginary unit
+            constexpr amrex::Real c = PhysConst::c;
+            constexpr amrex::Real c2 = c*c;
+            constexpr amrex::Real mu0 = PhysConst::mu0;
+            constexpr Complex I = Complex{0._rt, 1._rt};
+
+            const amrex::Real kx2 = kx*kx;
+            const amrex::Real ky2 = ky*ky;
+            const amrex::Real kz2 = kz*kz;
+            const amrex::Real knorm = std::sqrt(kx2 + ky2 + kz2);
+            const amrex::Real knorm2 = knorm*knorm;
+            const amrex::Real knorm4 = knorm2*knorm2;
+
+            const amrex::Real C = std::cos(c*knorm*dt);
+            const amrex::Real S = std::sin(c*knorm*dt);
+
+            // Update equations
+
+            if (knorm == 0._rt)
+            {
+                fields(i,j,k,Idx.Ex) = Ex_old - mu0*c2*dt*Jx_c0 - 0.5_rt*mu0*c2*dt2*Jx_c1;
+                fields(i,j,k,Idx.Ey) = Ey_old - mu0*c2*dt*Jy_c0 - 0.5_rt*mu0*c2*dt2*Jy_c1;
+                fields(i,j,k,Idx.Ez) = Ez_old - mu0*c2*dt*Jz_c0 - 0.5_rt*mu0*c2*dt2*Jz_c1;
+
+                if (div_cleaning)
+                {
+                    fields(i,j,k,Idx.F) = F_old - mu0*c2*dt*rho_c0 - 0.5_rt*mu0*c2*dt2*rho_c1;
+                }
+            }
+            else // knorm != 0
+            {
+                Complex C01, C02, C03, C04, C05, C06, C07, C08,
+                        C09, C10, C11, C12, C13, C14, C15, C16;
+
```

Comment:
We might want to define `1/knorm2`, `1/knorm` and `1/knorm4` as intermediate variables, to save some computation.

Reply:
Done in f821d07.